framework/db/ActiveRecord.php

```diff
@@ -291,12 +303,18 @@ public static function tableName()
      */
     public static function getTableSchema()
     {
-        $schema = static::getDb()->getSchema()->getTableSchema(static::tableName());
-        if ($schema !== null) {
-            return $schema;
-        } else {
-            throw new InvalidConfigException("The table does not exist: " . static::tableName());
+        static $tableSchema = null;
+
+        if ($tableSchema === null) {
+            $schema = static::getDb()->getSchema()->getTableSchema(static::tableName());
```

Comment:
Right, of course. You can still refresh the schema in Schema, but you can't propagate it to any already instanciated class.

We could fix this by adding the $refresh parameter to ActiveRecord::getTableSchema() as well, like this.

```php
public static function getTableSchema($refresh = false)
{
    static $tableSchema = null;
    
    if ($refresh === true) {
        $tableSchema = null;
    }

    if ($tableSchema === null) {
        $schema = static::getDb()->getSchema()->getTableSchema(static::tableName(), $refresh);
        if ($schema !== null) {
            $tableSchema = $schema;
        } else {
            throw new InvalidConfigException("The table does not exist: " . static::tableName());
        }
    }

    return $tableSchema;
}
```

I understand that the schema data is already cached in Schema::getTableSchema(), but even then ActiveRecord::getTableSchema() is called very frequently and as you can see in my profiling results the overhead of all these small function calls does add up, while for the vast majority of applications a class' schema would never change in the middle of a request.